File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.java
import java.io.*;
import java.util.Stack;

public class Main{
    public static void main(String[] args) throws IOException{

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));

        while (true) {

            String input = br.readLine();

            if (input.equals(".")) {
                break;
            }

            Stack<Character> stack = new Stack<>();
            Boolean flag = true;
            for (char ch : input.toCharArray()) {

                if (ch == '(' || ch == '[') {
                    stack.add(ch);
                } else if (ch == ')' || ch ==']') {

                    if (stack.isEmpty()) {
                        flag = false;
                        break;
                    }
                    char c = stack.pop();

                    if ((ch == ']' && c != '[') || (ch == ')' && c != '(')) {
                        flag = false;
                        break;
                    }
                }
            }
            
            if (stack.isEmpty() && flag) {
                bw.write("yes\n");
            } else {
                bw.write("no\n");
            }
        }
        bw.flush();
    }
}

File: 백준/Silver/4949. 균형잡힌 세상/균형잡힌 세상.py
import sys
from collections import deque

if __name__ == '__main__':

    ans = []
    while True:
        inputs = sys.stdin.readline().replace("\n","")

        if inputs == ".":
            break

        else:
            stack = deque()
            flag = True
            for ch in inputs:
                if ch == '(' or ch == '[':
                    stack.append(ch)

                elif ch == ')' or ch == ']':

                    # 비어 있으면 false
                    if not stack:
                        flag = False
                        break

                    c = stack.pop()

                    if (ch == ')' and c == '(') or (ch == ']' and c == '['):
                        continue
                    else:
                        flag = False

            if flag and not stack:
                ans.append("yes")
            else:
                ans.append("no")

    print("\n".join(ans))